Source/JavaScriptCore/runtime/PropertyTable.h

```diff
@@ -238,182 +213,240 @@ class PropertyTable final : public JSCell {
 
     static constexpr unsigned MinimumTableSize = 16;
+    static_assert(MinimumTableSize >= 16, "compact index is uint8_t and we should keep 16 byte aligned entries after this array");
 };
 
-inline PropertyTable::iterator PropertyTable::begin()
-{
-    auto* tableEnd = this->tableEnd();
-    return iterator(skipDeletedEntries(table(), tableEnd), tableEnd);
-}
-
-inline PropertyTable::iterator PropertyTable::end()
-{
-    auto* tableEnd = this->tableEnd();
-    return iterator(tableEnd, tableEnd);
-}
-
-inline PropertyTable::const_iterator PropertyTable::begin() const
-{
-    auto* tableEnd = this->tableEnd();
-    return const_iterator(skipDeletedEntries(table(), tableEnd), tableEnd);
-}
-
-inline PropertyTable::const_iterator PropertyTable::end() const
-{
-    auto* tableEnd = this->tableEnd();
-    return const_iterator(tableEnd, tableEnd);
-}
-
-inline PropertyTable::find_iterator PropertyTable::find(const KeyType& key)
+template<typename Index, typename Entry>
+std::tuple<unsigned, unsigned, PropertyOffset, uint8_t> PropertyTable::findImpl(const Index* indexVector, const Entry* table, const KeyType& key)
 {
-    ASSERT(key);
-    ASSERT(key->isAtom() || key->isSymbol());
     unsigned hash = IdentifierRepHash::hash(key);
 
 #if DUMP_PROPERTYMAP_STATS
     ++propertyTableStats->numFinds;
 #endif
 
     while (true) {
-        unsigned entryIndex = m_index[hash & m_indexMask];
+        unsigned index = hash & m_indexMask;
+        unsigned entryIndex = indexVector[index];
         if (entryIndex == EmptyEntryIndex)
-            return std::make_pair((ValueType*)nullptr, hash & m_indexMask);
-        if (key == table()[entryIndex - 1].key)
-            return std::make_pair(&table()[entryIndex - 1], hash & m_indexMask);
+            return std::tuple { entryIndex, index, invalidOffset, 0 };
+        const auto& entry = table[entryIndex - 1];
+        if (key == entry.key()) {
+            ASSERT(!m_deletedOffsets || !m_deletedOffsets->contains(entry.offset()));
+            return std::tuple { entryIndex, index, entry.offset(), entry.attributes() };
+        }
 
 #if DUMP_PROPERTYMAP_STATS
         ++propertyTableStats->numCollisions;
 #endif
 
 #if DUMP_PROPERTYMAP_COLLISIONS
         dataLog("PropertyTable collision for ", key, " (", hash, ")\n");
-        dataLog("Collided with ", table()[entryIndex - 1].key, "(", IdentifierRepHash::hash(table()[entryIndex - 1].key), ")\n");
+        dataLog("Collided with ", entry.key(), "(", IdentifierRepHash::hash(entry.key()), ")\n");
 #endif
 
         hash++;
     }
 }
 
-inline PropertyTable::ValueType* PropertyTable::get(const KeyType& key)
+inline std::tuple<unsigned, unsigned, PropertyOffset, uint8_t> PropertyTable::find(const KeyType& key)
+{
+    ASSERT(key);
+    ASSERT(key->isAtom() || key->isSymbol());
+    uintptr_t indexVector = m_indexVector;
+    if (indexVector & isCompactFlag) {
+        const auto* index = bitwise_cast<const uint8_t*>(indexVector & indexVectorMask);
```

Comment:
Might even be worth adding a template parameter to `indexedVector` getter I suggested. By default, it's `void*`, but it can be passed `uint8_t*` and `uint32_t*`, etc